src/common/mfu_io.c

```diff
@@ -603,6 +1197,51 @@ int mfu_closedir(DIR* dirp)
     return rc;
 }
 
+int mfu_file_closedir(DIR* dirp, mfu_file_t* mfu_file)
+{
+    if (mfu_file->type == POSIX) {
+        int rc = mfu_closedir(dirp);
+        return rc;
+    } else if (mfu_file->type == DAOS) {
+        int rc = daos_closedir(dirp, mfu_file);
+        return rc;
+    } else {
+        MFU_ABORT(-1, "File type not known: errno=%d (%s)",
+                  errno, strerror(errno));
+    }
+}
+
+struct dirent* daos_readdir(DIR* _dirp, mfu_file_t* mfu_file)
+{
+#ifdef DAOS_SUPPORT
+    struct dfs_mfu_t *dirp = (struct dfs_mfu_t *)_dirp;
+    if (dirp->num_ents) {
+        goto ret;
+    }
+    dirp->num_ents = NUM_DIRENTS;
+    int rc;
+    while (!daos_anchor_is_eof(&dirp->anchor)) {
+        rc = dfs_readdir(mfu_file->dfs, dirp->dir,
+                         &dirp->anchor, &dirp->num_ents,
+	                 dirp->ents);
+        if (rc) {
+            MFU_LOG(MFU_LOG_ERR, "dfs_readdir failed (%d %s)", rc, strerror(rc));
+            errno = ENOENT;
```

Comment:
In the case of an error, do you know whether dfs_readdir will set ```dirp->num_ents=0``` here.  If not, let's add a line here to do that.  That will fail more gracefully in the case where someone does this:
```
dirent* dent = daos_readdir(dirp, mfu_file);

// dfs_readdir returns an error, so daos_readdir sets errno and returns NULL

// user calls daos_readdir again
dent = daos_readdir(dirp, mfu_file);
```
If dirp->num_ents is not set to 0 on that first error, I think the second call will not behave correctly, since it will try to execute these lines which is probably not valid:
```
ret:
    dirp->num_ents--;
    return &dirp->ents[dirp->num_ents];
```